lib/github/kv.rb

```diff
@@ -252,6 +253,115 @@ def setnx(key, value, expires: nil)
+      validate_key(key)
+      validate_amount(amount) if amount
+      validate_expires(expires) if expires
+      validate_touch(touch_on_insert, expires)
+
+      expires ||= GitHub::SQL::NULL
+
+      # This query uses a few MySQL "hacks" to ensure that the incrementing
+      # is done atomically and the value is returned. The first trick is done
+      # using the `LAST_INSERT_ID` function. This allows us to manually set
+      # the LAST_INSERT_ID returned by the query. Here we are able to set it
+      # to the new value when an increment takes place, essentially allowing us
+      # to do: `UPDATE...;SELECT value from key_value where key=:key` in a
+      # single step.
+      #
+      # However the `LAST_INSERT_ID` trick is only used when the value is
+      # updated. Upon a fresh insert we know the amount is going to be set
+      # to the amount specified.
+      #
+      # Lastly we only do these tricks when the value at the key is and integer.
+      # If the value is not an integer the update ensures the values remain the
+      # same and we raise an error.
+      encapsulate_error {
+        sql = GitHub::SQL.run(<<-SQL, key: key, amount: amount, now: now, expires: expires, touch: !touch_on_insert, connection: connection)
+          INSERT INTO key_values (`key`, `value`, `created_at`, `updated_at`, `expires_at`)
+          VALUES(:key, :amount, :now, :now, :expires)
+          ON DUPLICATE KEY UPDATE
+            `value`=IF(
+              concat('',`value`*1) = `value`,
+              LAST_INSERT_ID(IF(
+                `expires_at` IS NULL OR `expires_at`>=:now,
+                `value`+:amount,
+                :amount
+              )),
+              `value`
+            ),
+            `updated_at`=IF(
+              concat('',`value`*1) = `value`,
+              :now,
+              `updated_at`
+            ),
+            `expires_at`=IF(
+              concat('',`value`*1) = `value`,
+              IF(
+                :touch,
+                :expires,
+                `expires_at`
+              ),
+              `expires_at`
+            )
+        SQL
+
+        # Check if MySQL connection parameter CLIENT_FOUND_ROWS is set. If it is
+        # to check for invalid values we must use a workaround via the last_insert_id
+        # From the MySQL docs:
+        #
+        # If you specify the CLIENT_FOUND_ROWS flag to the mysql_real_connect() C API
+        # function when connecting to mysqld, the affected-rows value is 1 (not 0)
+        # if an existing row is set to its current values.
+        # https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html
+        flags = connection.raw_connection.query_options[:flags]
+
+        # The flag is Mysql2::Client::FOUND_ROWS however older versions of the Mysql2
+        # library don't explicitly define the FOUND_ROWS const. The value of the const
+        # is 1<<1 (2)
+        check = 1 << 1
+        found_rows_is_set = flags & check == check
```

Comment:
The desirable configuration is to not have the flag set. Unfortunately GitHub does have it set (for historical reasons). Since we can't change the flag for GitHub I've introduced this workaround mostly for us. However the desirable and simple way to check is if the flag is not set. The only difference would be removing the first part of the `||` clause.

The workaround has one minor flaw though. If an insert takes place and the ID of the row is 0 then we'll throw an `InvalidValueError` I can't find any way around this. However the DBMS will start auto-incrment rows at 1 so this _shouldn't_ be an issue unless the DBMS is specifically told to increment the row ID starting at 0. This is the main reason having the flag not set is preferred.

Reply:
Actually now that I think about it we would need both OR we would need to come up with another workaround. Since the workaround relies on `sql.affected_rows == 1` but that would not be the case if the flag wasn't set. I think we need to keep it this way.

Reply:
Yeah. This is complicated. Hah. We should definitely be running tests with both configurations.

Reply:
> So, you're saying that GitHub has the flag set and that's the default for ActiveRecord and it's hard for people to disable manually? If that's the case, let's just raise an error if #found_rows_is_set? is false. This library is only meant to work with MySQL and ActiveRecord, so I don't think it would be problematic to raise an error if someone's gone to great lengths to unset this flag.

Is KV based on active record or MySQL2? I think the MySQL2 gem has it disabled by default and ActiveRecord enables it. I thought KV just took a MySQL2 connection.